src/storage-controller/src/lib.rs

```diff
@@ -1916,6 +1922,84 @@ where
         let id = self.introspection_ids[&type_];
         self.append_to_managed_collection(id, updates).await;
     }
+
+    /// With the CRDB based timestamp oracle, there is no longer write timestamp
+    /// fencing. As in, when a new Coordinator, `B`, starts up, there is nothing
+    /// that prevents an old Coordinator, `A`, from getting a new write
+    /// timestamp that is higher than `B`'s boot timestamp. Below is the
+    /// implications for all persist transaction scenarios, `on` means a
+    /// Coordinator turning the persist txn flag on, `off` means a Coordinator
+    /// turning the persist txn flag off.
+    /// - `off` -> `off`: The following series of events seem possible:
+    ///   1. `A` writes at `t_0`, s.t. `t_0` > `B`'s boot timestamp.
+    ///   2. `B` writes at `t_1`, s.t. `t_1` > `t_0`.
+    ///   3. `A` writes at `t_2`, s.t. `t_2` > `t_1`.
+    ///   4. etc.
+    /// - `on` -> `on`: The txn-shard is meant to correctly handle two writers
+    ///   so this should be fine. Note it's possible that we have two
+    ///   Coordinators interleaving write transactions without the leadership
+    ///   check described below, but that should be fine.
+    /// - `off` -> `on`: If `A` gets a write timestamp higher than `B`'s boot
+    ///   timestamp, then `A` can write directly to a data shard after it's been
+    ///   registered with a txn-shard, breaking the invariant that no data shard
+    ///   is written to directly while it's registered to a transaction shard.
+    ///   To mitigate this, we should do a leadership check AFTER getting the
+    ///   write timestamp. In order for `B` to register a data shard in the txn
+    ///   shard, it must first become the leader then second get a register
+    ///   timestamp. So if `A` gets a write timestamp higher than `B`'s register
+    ///   timestamp, it will fail the leadership check.
+    /// - `on` -> `off`: If `A` tries to write to the txn-shard at a timestamp
+    ///   higher than `B`'s boot timestamp, it will fail because the shards have
+    ///   been forgotten. So everything should be ok.
+    ///
+    ///  In general, all transitions make the following steps:
+    ///   1. Get write timestamp, `ts`.
+    ///   2. Apply all transactions to all data shards up to `ts`.
```

Comment:
Do we want to add a step after getting the write timestamp for the leadership check?

Reply:
This is mostly what happens on boot, which doesn't do the leadership check, no? Maybe I'm missing something about your original intent for this bit

Or is your point that we also need the leadership check for the init_ts? Hmmm...

Reply:
Oh I think I was just misreading this, I think we don't need a leadership check during boot.